### PR TITLE
docs(all): use remark-validate-links to check internal links, convert to file-path links

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This adds a gh actions workflow which runs [this action](https://github.com/gaurav-nelson/github-action-markdown-link-check) which in turns runs [this CLI program](https://github.com/tcort/markdown-link-check) which will check links in Markdown files in order to see if any are returning 404s.

It also checks that `#anchor` links have a matching anchor in the same file, which is pretty neat! So if you have some markdown like this:

```md
# my anchor

And also see [my anchor](#my-anchovy)
```

you'll get an error! neat!

Between this program and the internal link checking that Docusaurus already does we have a fairly robust 'test suite' here for our markdown.

### CODE REVIEW

File: mlc_config.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "ignorePatterns": [
+    { "pattern": "^/" },
+    { "pattern": "^./" },
+    { "pattern": "^../" }
+  ],
+  "timeout": "20s"
+}
```
Comment (alicewriteswrongs) [inline mlc_config.json:8]: this is the configuration for the [markdown link check](https://github.com/tcort/markdown-link-check) program.

- the timeout is 20s because we have a few archive.org links which can be erroneously marked as dead because they take a while to respond
- the `ignorePatterns` are set to `/`, `./` and `../` because that lets us have the markdown link checker ignore all the _internal_ links, the checking of which is left to Docusaurus (it does a pretty good job already)

One consequence of that `ignorePatterns` bit that you'll notice is that it requires us to always use `./` or `../` for internal links in order to ensure that the link checker doesn't report spuriously broken links.

basically, if it encounters

```md
link to [some page](dir/page)
```

it will try to resolve `dir/page` as a URL, since it doesn't understand Docusaurus' format and algorithm for turning that string into a link to, say, `/dir/page`. 

So we just need to use the `./dir/page` path instead, which is, as far as I can tell, equally supported by Docusaurus.

**Review by alicewriteswrongs — COMMENTED** [review]
noting one thing

File: docs/components/properties.md [hunk 2]
```diff
@@ -645,7 +645,7 @@ export class ToDoListItem {
 
 ## Prop Validation
 
-To do validation of a Prop, you can use the [@Watch()](reactive-data/#watch-decorator) decorator:
+To do validation of a Prop, you can use the [@Watch()](./reactive-data/#watch-decorator) decorator:
```
Comment (rwaskiewicz) [inline docs/components/properties.md:648]: Does this link exist? I see `docs/reactive-data/#the-watch-decorator-watch`, but not what's listed here

File: mlc_config.json (hunk 1 above)
Comment (rwaskiewicz) [reply mlc_config.json:8]: I love it - I hate this config uses JSON (not that it's our choice) and we can't add this as comments :-(

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
It looks like there's some false positives with respect to links with anchors in them 😬 I'm a little concerned about adding something that _says_ everything's OK, but technically has broken links. Perhaps this is better than nothing though 🤔

File: docs/components/api.md [hunk 3]
```diff
@@ -23,28 +23,28 @@ The whole API provided by stencil can be condensed in a set of decorators, lifec
...
-- [@Component()](component#component-decorator) declares a new web component
-- [@Prop()](properties#prop-decorator) declares an exposed property/attribute
-- [@State()](state#the-state-decorator-state) declares an internal state of the component
-- [@Watch()](reactive-data#watch-decorator) declares a hook that runs when a property or state changes
-- [@Element()](host-element#element-decorator) declares a reference to the host element
-- [@Method()](methods#method-decorator) declares an exposed public method
-- [@Event()](events#event-decorator) declares a DOM event the component might emit
-- [@Listen()](events#listen-decorator) listens for DOM events
+- [@Component()](./component#component-decorator) declares a new web component
+- [@Prop()](./properties#prop-decorator) declares an exposed property/attribute
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply docs/components/api.md:27]: It may not, I think it will unfortunately only check anchors within the same file. What I ended up doing was basically converting all of these links to be relative links that start with `./`. This works with Docusaurus properly, and it will check that the link resolves to a page, but I don't think that Docusaurus verifies that specific anchors within a linked page exist.

The markdown link checker, by contrast, checks that links either resolve to anchor links *within the same file* (if they look like `[foo](#foo)` _or_ that they'll return a 200 when it makes a `GET` request to them (more or less). So basically it won't know what to do with the links that Docusaurus renders as internal links.

So anyhow, what I did to get Docusaurus' link-checking and the markdown link checker's link checking to work somewhat well together is converting Docusaurus links to use the relative path and then ignoring those links in the markdown link checker.

However, one thing I just tried is using the alternate file path linking supported in Docusaurus, like

```md
[link](./foo.md#foo)
```

Both Docusaurus and the markdown link checker will check this and report that yes, this is pointing to a file in the right spot, but neither will check that the anchor tag `foo` is present in that file. Boo!

There's an issue on the Docusaurus repo here about this: https://github.com/facebook/docusaurus/issues/3321

One suggestion in there is to convert to file-path links (`[](./foo.md)`) and then use a project like this one: https://github.com/remarkjs/remark-validate-links which might provide a better way to validate all the _internal_ links between files. It looks like it checks that anchor tags are present in any linked file.

_That_ option of course doesn't cover _external_ links.

So my takeaway here is that there's not a perfect solution here. I think I'd be in favor of the following:

- adding the markdown link checker, along the lines of how it's set up here where it's constrained to only checking external URLs
- refactoring internal links to use the filepath-based linking
- adding `remark-validate-links` to check that all internal links + anchor tags are valid (this would make it basically an extra layer of validation on top of what will already get checked by `docusaurus build`).

Thoughts?

[the remark this answers is not in the document]
Comment (tanner-reits) [reply docs/components/api.md:27]: The remark plugin seems pretty cool as long as all of the tools can play together nicely. Writing relative file paths might be a bit tedious, but I feel like we had to do that sometimes in the old docs too so nbd. I don't really have a strong opinion here for next steps, but think this is a step in the right direction and probably better than the nothing we have today 😬

[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply docs/components/api.md:27]: This sounds like a plan to me. The only thing I'm hesitant on is converting internal links to use the filepath-based linking. I think this could make some maintenance a little tricky if we rename a file, but perhaps that's the cost you have to pay if you want anchor tag validation at the moment. I think I'm willing to pay that cost if it means automated tests. Does `remark-validate-links` flag you if you don't use file-path links? My other concern here is education across teams contributing to the site. If `remark-validate-links` doesn't flag non-file-path links, maybe the compromise here is to just move forward with `mlc` only here, and see how that works for us (taking that usage period as a time to better understand our needs)

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply docs/components/api.md:27]: So I was just able to confirm that `remark-validate-links` _does_ show a warning if you don't use file-path-based links. If you have, for instance, something like this:

```md
| `single-export-module`        | All component and custom element definition helper functions will be exported from the `index.js` file in the output directory. This file can be used as the root module when distributing your component library, see [below](../custom-elements#distributing-custom-elements) for more details. |
```

it will say this:

Screenshot 2023-02-09 at 10 33 40 AM

this can be addressed (in this case) by changing the markdown to this:

```md
| `single-export-module`        | All component and custom element definition helper functions will be exported from the `index.js` file in the output directory. This file can be used as the root module when distributing your component library, see [below](../output-targets/custom-elements.md#distributing-custom-elements) for more details. |
```

A slightly confusing thing with re-writing these to relative-file-path based links is that right now they are effectively relative paths _based on their final url_, so for instance from `docs/introduction/upgrading-to-stencil-three.md`, which has the `url` set to `url: /docs/upgrading-to-stencil-3` in the frontmatter the relative url to the page rendered based on `docs/output-targets/custom-elements.md` is actually just `../custom-elements` because _that_ file has this in frontmatter: `slug: /custom-elements`

So it's a little bit funky to rewrite them, you can sometimes just put `.md` at the end of what's there already, but sometimes you have to figure out which `.md` file the relative URL would normally point to and then find it. Not impossible, but a little funky.

One consequence of that is that right now individual markdown files can be somewhat 'mobile', i.e. they can be moved around without all their relative links being broken. With this change, moving a file would require fixing all of the relative paths to other files (possibly some IDEs can fix such things easily). Anyway, not a major concern because we don't move files around all that much, but worth flagging.

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply docs/components/api.md]: I just pushed a change to convert all the internal links over to filepath based ones, fixing any missing anchor tags as well. I used `remark-validate-links` locally to do so, it's quite handy! If you have `[](./foo.md#foo-bar)` and it should actually be `[](./foo.md#foo-bar-baz)` it will do a little 'did you mean.. ?' suggestion (I think based on prefix matching). It's not so many that need to change actually, and it did catch a few broken anchors, so that's nice.

The question then is, if we want to use the `remark-validate-links` thing, should that be installed as a dependency, or installed only as a github action? I'd kind of lean towards installing as a project dependency so we can add a `package.json` script for running it for checking things out locally before pushing.

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply docs/components/api.md]: ok so update — that only addressed a small portion of the internal links. There's a lot more!

File: docs/README.md [hunk 4]
```diff
@@ -1,11 +1,11 @@
 # Table of Contents
```
[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply docs/README.md:1]: Created https://github.com/ionic-team/stencil-site/pull/1037 while waiting for the Stencil nightly CI fix to finish running

File: docs/components/api.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply docs/components/api.md]: 👍 Would you consider this ready to re-review then? I'd like to pull the branch down and do some faux-development with it.

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply docs/components/api.md]: feel free to play with it, I am going through and changing over to relative links everywhere -- this requires editing ~400 or so of them and I'm not quite done yet!

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply docs/components/api.md]: I'm going to add `remark-validate-links` to dev dependencies and then add running it to the CI as well

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz @tanner-reits Ok so I've made some changes on here. These are the key changes:

- added the remark linting thingy for checking internal links, including checking that headers exist in linked-to files
- converted a fairly large number of links to be relative, file-path-based links
- added that internal link checker to the CI

As of right now there is more to do in order to convert all of the links to file-path links, which is what would give us complete checking of all of them (including headers). We could either do that here or do some subsequent PRs in order to make those changes, what do you two think?

**Comment by rwaskiewicz** [comment]
@alicewriteswrongs I say we just do it all here if you think the effort is relatively well sized. Unsure if we need to do _all_ of it now to get CI to pass or not. Not that that precludes us from splitting the effort up. I think my vote would be if the effort is small-medium sized we do it here. Otherwise maybe it makes sense to split this PR up and the subsequent work as well.

**Comment by alicewriteswrongs** [comment]
alright I think this should be ready for another look. I went through and updated all of the internal links to be relative file path links. The process I followed to do this was roughly like the following:

- use `sed` to turn all links of the form `[text](/url)` to `[text](./url)` so that they would start creating errors when running `npm run check-links`
- open up the files showing errors in the output of that command one-by-one and fix all the errors. to fix errors I:
    - looked at what the original url was (e.g. `/assets`)
    - grep for the file with the `slug` that sets that url (e.g. `grep 'slug: /assets'`)
    - update the link to have the right relative path to that file (e.g. `../guides/assets.md`)

I additionally made corrections if there were any non-existent headers encountered, e.g. if there was a link like `[dist](./dist#header-that-doesnt-exist)` I would try to find the right header or, if none was obvious, I would just remove it.

Now all the internal links are passing the `check-links` command, and I've read through the diff and I believe that they all point to the right place.

File: versioned_docs/version-v3.0/components/properties.md [hunk 5]
```diff
@@ -645,7 +645,7 @@ export class ToDoListItem {
 
 ## Prop Validation
 
-To do validation of a Prop, you can use the [@Watch()](reactive-data/#watch-decorator) decorator:
+To do validation of a Prop, you can use the [@Watch()](./reactive-data.md#the-watch-decorator-watch) decorator:
```
Comment (alicewriteswrongs) [inline versioned_docs/version-v3.0/components/properties.md:648]: Here's an example of a nonexistent header that the link checker caught

**Review by alicewriteswrongs — COMMENTED** [review]
one little note

File: docs/guides/assets.md [hunk 6]
```diff
@@ -26,11 +26,11 @@ The **asset base path** is the directory that Stencil will use to resolve assets
 When a component uses an asset, the asset's location is resolved relative to the asset base path.
 
 The asset base path is automatically set for the following output targets:
-- [dist](/distribution)
-- [hydrate](/hydrate-app)
-- [www](/www)
+- [dist](../output-targets/dist.md)
+- [hydrate](./hydrate-app.md)
```
Comment (rwaskiewicz) [inline docs/guides/assets.md:30]: Non blocking: This URL is different from what we use in the [v2](https://github.com/ionic-team/stencil-site/pull/1022/files#diff-419f03d30bdf123f841df6e0dfad6010ae1f847f25d3511889bbe212d40f1fbeR30) & [v3.0 docs](https://github.com/ionic-team/stencil-site/pull/1022/files#diff-419f03d30bdf123f841df6e0dfad6010ae1f847f25d3511889bbe212d40f1fbeR30). They resolve to the same place, just a minor difference in their values. Calling this out just in case that causes alarms to go off in your head/there's an issue I'm not seeing.

File: versioned_docs/version-v3.0/guides/assets.md [hunk 7]
```diff
@@ -128,19 +128,19 @@ export class MyComponent {
...
 ### Manually Moving Assets
 
-For the [dist-custom-elements](/custom-elements) output target, options like `assetsDirs` do not copy assets to the distribution directory.
+For the [dist-custom-elements](../output-targets/custom-elements.md) output target, options like `assetsDirs` do not copy assets to the distribution directory.
 
 It's recommended that a bundler (such as rollup) or a Stencil `copy` task is used to ensure the static assets are copied to the distribution directory.
 
 #### Stencil Copy Task
 
-[Stencil `copy` task](/copy-tasks)'s can be used to define files and folders to be copied over to the distribution directory.
```
Comment (rwaskiewicz) [inline versioned_docs/version-v3.0/guides/assets.md:143]: Minor, I think we dropped an apostrophe here that we didn't lose in other docs

File: versioned_docs/version-v3.0/output-targets/custom-elements.md [hunk 8]
```diff
@@ -14,7 +14,7 @@ contributors:
 
 The `dist-custom-elements` output target creates custom elements that directly extend `HTMLElement` and provides simple utility functions for easily defining these elements on the [Custom Element Registry](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry). This output target excels in use in frontend frameworks and projects that will handle bundling, lazy-loading, and custom element registration themselves.
 
-This target can be used outside of frameworks as well, if lazy-loading functionality is not required or desired. For using lazily loaded Stencil components, please refer to the [dist output target](/distribution).
+This target can be used outside of frameworks as well, if lazy-loading functionality is not required or desired. For using lazily loaded Stencil components, please refer to the [dist output target](../output-targets/dist.md).
```
Comment (rwaskiewicz) [inline versioned_docs/version-v3.0/output-targets/custom-elements.md:17]: Another (non-blocking) case where the link differs between the versions of the docs - see https://github.com/ionic-team/stencil-site/pull/1022/files#diff-009a946cd7a1bed32d997de1a2f768193f39f0285964f590a32c392fa5cd3d15R17

File: versioned_docs/version-v2/static-site-generation/01-overview.md [hunk 9]
```diff
@@ -31,7 +31,7 @@ Stencil makes SSG easy, so read on to see how to incorporate it into your apps.
 
 ## How Static Site Generation and Prerendering Works
 
-**Build Hydrate App**: The first step in prerendering is for the compiler to generate a "hydrate" app, which is a single directory to be used by Node.js. The "hydrate" app is automatically generated when the `--prerender` CLI flag is provided and by default the app is saved to `dist/hydrate`. Prerendering uses the hydrate app internally, however it can be used directly at a lower-level. [Learn more about the Hydrate App](./hydrate-app).
```
Comment (rwaskiewicz) [inline versioned_docs/version-v2/static-site-generation/01-overview.md:34]: I think we accidentally dropped a '.' here

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
All in all looks good! 

I verified all the links for v3.0, then generated a diff file for each directory containing versioned docs (e.g. `git diff main versioned_docs/version-v2 > v2.0-diff`, then also for v3 & next), then diffed those diff files to root out any tiny subtle changes between the versions. I caught a few things like dropped periods/apostrophes, but all this is really solid!

In addition to my comments inlined, can we change the name of this PR to:
- use `docs():` nomenclature
- update the name, since we're not using MLC at the moment?

**Review by tanner-reits — COMMENTED** [review]
Nice job! I validated the `check-links` command and ran through the diffs. Lot to parse through, but didn't see anything glaring. Looks like there are conflicts with the updated React integration docs though

File: docs/guides/assets.md (hunk 6 above)
Comment (alicewriteswrongs) [reply docs/guides/assets.md:30]: ah yeah I think that's fine, just a side-effect of trying to do these edits quickly haha

File: versioned_docs/version-v3.0/guides/assets.md (hunk 7 above)
Comment (alicewriteswrongs) [reply versioned_docs/version-v3.0/guides/assets.md:143]: ah yeah, I think we want to remove that apostrophe - I'll remove it on the other files too

File: mlc_config.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
+{
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply mlc_config.json]: no longer needed! I deleted it

**Comment by alicewriteswrongs** [comment]
ok I think those issues should be addressed, ready for a re-review

**Review by rwaskiewicz — APPROVED** [review]

**Review by tanner-reits — APPROVED** [review]
